### ISSUE TITLE
Kotlin Multiplatform Library

### ISSUE THREAD

**Issue by Legion2** [PostHog/posthog-android#8]
Integrating Posthog in Kotlin multiplatform projects currently requires integrating the android and ios libraries separatly. With a multiplatform library the code for android and ios must not be duplicated and user must not themself create expect/actual declarations for PostHog. The kotlin library can also be used nativly in android only projects.

**Comment by Legion2** [comment]
We now internally maintain a basic multiplatform wrapper for posthog android and ios targets. If somebody is interested we can share it as open source.

**Comment by marandaneto** [comment]
We're currently reworking the Android SDK and it's rewritten in Kotlin, still has Java dependencies e.g. Gson but it'd be much easier to write a KMP wrapper on top of it.

Stay tuned.

**Comment by marandaneto** [comment]
A quick workaround, for now, is to just call the [API directly](https://posthog.com/tutorials/api-capture-events) until we provide first-class support for Kotlin Multiplatform.

**Comment by marandaneto** [comment]
@ar-g We are not focused on that right now since there's not a lot of demand, usually we prioritize depending on the number of requests, very user-centric.

Instead of writing a new SDK, it might be easier to start with a wrapper SDK similar to https://github.com/postHog/posthog-flutter which just forwards the call to the Native SDKs (Android, iOS, and the JS SDK).

Would that help? Would you be willing to submit a PR? I'm happy to guide you and review it, let me know if that would be an alternative.

**Comment by ar-g** [comment]
@marandaneto our end goal is to use it in our own library that shipped to different clients, practically speaking we don't need any native bindings, screen tracking and such. Do we even need in that case to write a wrapper for other sdks, wouldn't it make more sense to implement own sender of events instead?

What we want to do is track transaction failure events that include versions and different metadata. 
We also want to set up PagerDuty based on the threshold of these failures, is this possible? 
Another thing is that there're a lot of failure events that need to be filtered, this is probably should be doable using query language. Do you think PostHog is the right tool for this purpose? We use grafana for similar scenarios, and sentry for crash reports on platforms atm.

**Comment by marandaneto** [comment]
@ar-g most likely a pure client would make more sense, yes, more lightweight.

> What we want to do is track transaction failure events that include versions and different metadata.
That is doable with the [API](https://posthog.com/tutorials/api-capture-events), yes, if you don't need offline caching, that's easy.

> We also want to set up PagerDuty based on the threshold of these failures, is this possible?

Let me check that.

> Another thing is that there are a lot of failure events that need to be filtered, this is probably should be doable using query language

That is correct, filtering is fine.

> Do you think PostHog is the right tool for this purpose? We want to achieve something similar we have with grafana and backend tracking atm.

I'd need to understand a bit more about what exactly you wanna track but dashboards and filtering should not be a problem.
If you're able to write your client library calling the API, all good, it should be super straightforward, you can probably borrow some code from the Android SDK which is Kotlin only.
About the alerting, will come back to you.

**Comment by marandaneto** [comment]
@ar-g about alerting, unfortunately not yet, please upvote this [issue](https://github.com/PostHog/posthog/issues/14331).

**Comment by ColtonIdle** [comment]
Did anyone get around to writing a kmp library/wrapper for posthog? we currently want to use posthog, but looks like theres no first class support.

**Comment by ColtonIdle** [comment]
is there some kind of way to vote for this support? or just commenting on it here?

**Comment by marandaneto** [comment]
> is there some kind of way to vote for this support? or just commenting on it here?

now you can https://posthog.com/roadmap?sort=team&team=Product+Analytics+Team
Vote for Kotlin Multiplatform SDK

**Comment by marandaneto** [comment]
@MamboBryan cool, I think a straightforward approach would be just to be a wrapper KMP SDK that forwards the call to the iOS and Android SDK, similar to our [Flutter SDK](https://github.com/PostHog/posthog-flutter/blob/main/lib/src/posthog_flutter_io.dart).
Most of the work would be around the tooling, getting the Gradle build scripts to do its work.

**Comment by MamboBryan** [comment]
> @MamboBryan cool, I think a straightforward approach would be just to be a wrapper KMP SDK that forwards the call to the iOS and Android SDK, similar to our [Flutter SDK](https://github.com/PostHog/posthog-flutter/blob/main/lib/src/posthog_flutter_io.dart).
> Most of the work would be around the tooling, getting the Gradle build scripts to do its work.

I was thinking the same thing. To get up and running we can just call the native ones directly 👍🏾. Should I do it on a private repo or you'll provide one?
Also since it's KMP are we saying we're looking at all the platforms it provides?

**Comment by MamboBryan** [comment]
> @MamboBryan any update on this?
Lemme try and work on it in the upcoming weeks. Thanks for the reminder @marandaneto

**Comment by iTomkinas** [comment]
Any updates?

**Comment by marandaneto** [comment]
TIL https://github.com/samuolis/posthog-kmp

**Comment by MohitMandalia** [comment]
@samuolis I would like to help on this, will be checking out your repo.

**Comment by griffinli** [comment]
Any update on official KMP support?

**Comment by marandaneto** [comment]
TIL https://github.com/AndroidPoet/posthog-kmp

**Comment by marandaneto** [comment]
Currently not prioritized for this Q.

**Comment by skaldebane** [comment]
It seems that things are finally moving in https://github.com/PostHog/posthog-kmp, based on @samuolis's fork!

**Comment by griffinli** [comment]
Finally!

**Comment by MohitMandalia** [comment]
> It seems that things are finally moving in https://github.com/PostHog/posthog-kmp, based on [@samuolis](https://github.com/samuolis)'s fork!

Thanks to @samuolis  we have things in good shape 🚀

**Comment by turnipdabeets** [comment]
Yes! Closing this issue for https://github.com/PostHog/posthog-kmp. Please bring any issue over there :)

**Comment by ColtonIdle** [comment]
Would be nice to have @samuolis in the acknowledgements section, no?
